### PR TITLE
sql,server: create toString to role option

### PR DESCRIPTION
**Pull request by maryliag**
Previously, the code was using roleOption.String() to
create a string to be used on queries. This was causing
an issue with roles that contained spaces within it,
e.g. role option `VALIDUNTIL` was being translate to
`VALIDUNTIL` instead of the correct way `VALID UNTIL`.
This commit updates the String() function to correctly
add the space on the `VALID UNTIL` case.

Release note (bug fix): Update the String() function of
roleOption to add a space on the role `VALID UNTIL`.

### CODE REVIEW

File: pkg/server/admin.go [hunk 1]
```diff
@@ -3209,7 +3209,7 @@ func (c *adminPrivilegeChecker) hasRoleOption(
 	row, err := c.ie.QueryRowEx(
 		ctx, "check-role-option", nil, /* txn */
 		sessiondata.InternalExecutorOverride{User: user},
-		"SELECT crdb_internal.has_role_option($1)", roleOption.String())
+		"SELECT crdb_internal.has_role_option($1)", roleoption.ToString[roleOption])
```
Comment (RichardJCai) [inline pkg/server/admin.go:3212]: nit: I'd prefer if we added a method to `RoleOption` called ToString() instead of using the map directly

**Review by rafiss — CHANGES_REQUESTED** [review]
can you add a test that would fail without this change?

**Review by maryliag — COMMENTED** [review]
Test added

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-:-MtyDpePAgcgBoUz-NnP:bdjrzdj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)

___
*[pkg/server/admin.go, line 3212 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MtxIC948eN-FR42nyMj-r1-3212:-MtyDvI1C1ghrkKeN5wy:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/6539edd55cb29aba36db1e19cea3dbf339049cf9/pkg/server/admin.go#L3212)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

nit: I'd prefer if we added a method to `RoleOption` called ToString() instead of using the map directly
</blockquote></details>

Done

**Review by matthewtodd — CHANGES_REQUESTED** [review]
> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75271)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag)

___
*[pkg/sql/roleoption/role_option.go, line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuBvLz1CgekbSNUBDap:-MuBvLz2206QuMrsI9jZ:b-p407bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a866f0c1d8254124101c6f609bdafa34ba28f3f6/pkg/sql/roleoption/role_option.go#L169)):*
> ```Go
> // It should be used instead of roleOption.String() to accommodate the cases
> // with space on the role (e.g. `VALID UNTIL`).
> func (o Option) ToString() string {
> ```

Looks like there's some precedence for the name `SQLString` here, WDYT?

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75271)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @matthewtodd)

___
*[pkg/sql/roleoption/role_option.go, line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuBvLz1CgekbSNUBDap:-MuBzJvU5gZQN5fHwx3T:b-2ts9r9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a866f0c1d8254124101c6f609bdafa34ba28f3f6/pkg/sql/roleoption/role_option.go#L169)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

Looks like there's some precedence for the name `SQLString` here, WDYT?
</blockquote></details>

Are you referring to the toSQLStmt mapping? I was thinking is just easier to understand/remember `ToString` than `ToSQLString`

**Review by matthewtodd — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75271)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag)

___
*[pkg/sql/roleoption/role_option.go, line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuBvLz1CgekbSNUBDap:-MuC01fK3z8ebaq6ahFN:bfj94tv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a866f0c1d8254124101c6f609bdafa34ba28f3f6/pkg/sql/roleoption/role_option.go#L169)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

Are you referring to the toSQLStmt mapping? I was thinking is just easier to understand/remember `ToString` than `ToSQLString`
</blockquote></details>

I was talking about renaming this method to `SQLString`. If you `git grep SQLString pkg` there are some examples. I was thinking I would have a hard time understanding the difference between `String` and `ToString` from the names alone.

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75271)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @matthewtodd)

___
*[pkg/sql/roleoption/role_option.go, line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuBvLz1CgekbSNUBDap:-MuC3Nw91_mxP5MFT9de:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a866f0c1d8254124101c6f609bdafa34ba28f3f6/pkg/sql/roleoption/role_option.go#L169)):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

I was talking about renaming this method to `SQLString`. If you `git grep SQLString pkg` there are some examples. I was thinking I would have a hard time understanding the difference between `String` and `ToString` from the names alone.
</blockquote></details>

Done

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75271)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @maryliag and @matthewtodd)

___
*[pkg/sql/roleoption/role_option.go, line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuBvLz1CgekbSNUBDap:-MuC66nh7n98MCPPl4Ns:b83s0lg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a866f0c1d8254124101c6f609bdafa34ba28f3f6/pkg/sql/roleoption/role_option.go#L169)):*
<details><summary><i>Previously, maryliag (Marylia Gutierrez) wrote…</i></summary><blockquote>

Done
</blockquote></details>

on the other hand, now i wonder if we ever want to use the `String()` method at all? should we just make `String()` be the only one?
___
*[pkg/sql/roleoption/role_option.go, line 170 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuC4Uve2JTJ4FcSo4TZ:-MuC4Uvf24G3VTgE8Hqf:b8a5uvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/352045512250f9ef9cac783486a7b1f5df822720/pkg/sql/roleoption/role_option.go#L170)):*
> ```Go
> // with space on the role (e.g. `VALID UNTIL`).
> func (o Option) SQLString() string {
> 	return toSQLStringMap[o]
> ```

is VALID UNTIL the only one that needs to be handled differently?

i wonder if we should just add a special case here in this function, instead of making a whole new map that needs to be maintained separately (and might be easy to forget to update)
___
*[pkg/server/user_test.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuC4eLv0qGBr1AfBufE:-MuC4eLv0qGBr1AfBufF:b-d8uleu) ([raw file](https://github.com/cockroachdb/cockroach/blob/352045512250f9ef9cac783486a7b1f5df822720/pkg/server/user_test.go#L41)):*
> ```Go
> 	for name := range roleoption.ByName {
> 		_, err := s.(*TestServer).status.baseStatusServer.privilegeChecker.hasRoleOption(ctx, fooUser, roleoption.ByName[name])
> 		require.NoError(t, err)
> ```

ah i didn't realize it was erroring before; i thought it was returning true/false incorrectly.

should we also add a test to make sure true/false is calculated correctly too? up to you

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75271)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @matthewtodd, @rafiss, and @RichardJCai)

___
*[pkg/sql/roleoption/role_option.go, line 169 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuBvLz1CgekbSNUBDap:-MuCOcfa69JG9NJ4zIy7:b-i660tk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a866f0c1d8254124101c6f609bdafa34ba28f3f6/pkg/sql/roleoption/role_option.go#L169)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

on the other hand, now i wonder if we ever want to use the `String()` method at all? should we just make `String()` be the only one?
</blockquote></details>

Changed the original String function,using the `//go:generate stringer -type=Option -linecomment`
Thanks Rafi for the tip!
___
*[pkg/sql/roleoption/role_option.go, line 170 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuC4Uve2JTJ4FcSo4TZ:-MuCOiX9Eiq8dS-9FAvB:bnwprfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/352045512250f9ef9cac783486a7b1f5df822720/pkg/sql/roleoption/role_option.go#L170)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is VALID UNTIL the only one that needs to be handled differently?

i wonder if we should just add a special case here in this function, instead of making a whole new map that needs to be maintained separately (and might be easy to forget to update)
</blockquote></details>

The only one different, so I added the comment on the role with the space
___
*[pkg/server/user_test.go, line 41 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuC4eLv0qGBr1AfBufE:-MuCOmxZ4VBXWu6HVwSa:be2ynim) ([raw file](https://github.com/cockroachdb/cockroach/blob/352045512250f9ef9cac783486a7b1f5df822720/pkg/server/user_test.go#L41)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ah i didn't realize it was erroring before; i thought it was returning true/false incorrectly.

should we also add a test to make sure true/false is calculated correctly too? up to you
</blockquote></details>

Add the check if the role existed after being added

**Review by matthewtodd — APPROVED** [review]
> Reviewed 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75271)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @maryliag and @rafiss)

___
*[-- commits, line 10 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuCPnky7vZXywVZCuZI):*
That new `//go:generate` line is awesome! Probably worth rewriting the commit message now? Otherwise :lgtm:

**Review by rafiss — APPROVED** [review]
this looks great! thanks for all the work on this

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-:-MuCQ7gn7WcsrC_TCnJ_:bk0ovhl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @maryliag and @rafiss)

___
*[pkg/server/user_test.go, line 62 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuCQ0Qa3pFYJ-htSlyi:-MuCQ0Qa3pFYJ-htSlyj:b5jcyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1da0247e16905275e7a6436d571935316512fa08/pkg/server/user_test.go#L62)):*
> ```Go
> 		require.NoError(t, err)
> 
> 		hasRole, err := s.(*TestServer).status.baseStatusServer.privilegeChecker.hasRoleOption(ctx, fooUser, roleoption.ByName[name])
> ```

this could also do a check using `crdb_internal.has_role_option` directly, but not necessary

**Review by maryliag — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75271)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @matthewtodd and @rafiss)

___
*[-- commits, line 10 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuCPnky7vZXywVZCuZI):*
<details><summary><i>Previously, matthewtodd (Matthew Todd) wrote…</i></summary><blockquote>

That new `//go:generate` line is awesome! Probably worth rewriting the commit message now? Otherwise :lgtm: 
</blockquote></details>

Updated
___
*[pkg/server/user_test.go, line 62 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/75271#-MuCQ0Qa3pFYJ-htSlyi:-MuCXd5X3YgLhzJ-rcGM:b-9uwpmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1da0247e16905275e7a6436d571935316512fa08/pkg/server/user_test.go#L62)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this could also do a check using `crdb_internal.has_role_option` directly, but not necessary
</blockquote></details>

I prefer keeping the hasRole function because it does a few extra checks and in the end that was the funcional that was causing issues for me and I wanted to test.

**Comment by maryliag** [comment]
TFTR!

bors r+

**Pull request merged**